### PR TITLE
feat: Add multisig script generation skeleton

### PR DESCRIPTION
**Pull request by classicalliu**
* I'm not sure should we use `R` and `M` or more readable `require_n` and `threshold`, former one is more friendly to who's familiar with `secp256k1_blake160_multisig`.
* Should we retain `fromAddress`, or we can delete this parameter and generate `fromScript` from `fromInfo(R, M, publicKeyHashes and so on...)` ?
* Should we expose `serializeMultisigScript` and `multisigArgs`, these will help to generate multisig address, or we can move to somewhere like `utils` ?

### CODE REVIEW

File: packages/common-scripts/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+   * @param fromAddress must be multisig script address
+   * @param fromInfo multisig parameters
+   * @param toAddress can be any type of lock script and can left empty
+   * @param amount transfer CKB capacity in shannon
+   * @param options
+   */
+  transfer(
+    txSkeleton: TransactionSkeleton,
+    fromAddress: string,
+    fromInfo: {
```
Comment (xxuejie) [inline packages/common-scripts/index.d.ts:19]: I think it is better to define a separate type here for fromInfo, we can also explain what R and M does in the type

File: packages/common-scripts/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      R: number,
+      M: number,
+      publicKeyHashes: string[],
+    },
+    toAddress: string | undefined,
```
Comment (xxuejie) [inline packages/common-scripts/index.d.ts:24]: Let's define an Address type which is just a string for now, later we can add more checking

File: packages/common-scripts/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    amount: string | bigint,
```
Comment (xxuejie) [inline packages/common-scripts/index.d.ts:25]: For now let's start strict: we only accept BigInt here

File: packages/common-scripts/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    options: {
+      config: any, // TODO: define this type later
```
Comment (xxuejie) [inline packages/common-scripts/index.d.ts:27]: We can define a Config type now which is just any, that will make future changes easier

File: packages/common-scripts/lib/secp256k1_blake160.js [hunk 5]
```diff
@@ -196,7 +195,7 @@ async function transfer(
     }
     let witness = txSkeleton.get("witnesses").get(firstIndex);
     const newWitnessArgs = {
-      /* 65 zeros in hex */
+      /* 65 bytes zeros in hex */
```
Comment (xxuejie) [inline packages/common-scripts/lib/secp256k1_blake160.js:198]: `65-byte` or `65 bytes'`

File: packages/common-scripts/lib/secp256k1_blake160_multisig.js [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    template.hash_type !== script.hash_type
+  ) {
+    throw new Error(
+      "Provided script is not SECP256K1_BLAKE160_MULTISIG script!"
+    );
+  }
+}
+
+/**
+ * secp256k1_blake160_multisig requires S, R, M, N and public key hashes
```
Comment (xxuejie) [inline packages/common-scripts/lib/secp256k1_blake160_multisig.js:31]: For now let's put all js/ts docs in TypeScript definition file.

File: packages/helpers/lib/index.js [hunk 7]
```diff
@@ -32,6 +34,8 @@ function minimalCellCapacity(fullCell, { validate = true } = {}) {
   let bytes = 8;
   bytes += new Reader(fullCell.cell_output.lock.code_hash).length();
   bytes += new Reader(fullCell.cell_output.lock.args).length();
+  // hash_type field
+  // hash_type field
```
Comment (xxuejie) [inline packages/helpers/lib/index.js:38]: Remove one line?

File: packages/common-scripts/index.d.ts (hunk 1 above)
Comment (classicalliu) [reply packages/common-scripts/index.d.ts]: Add an interface `MultisigScript`

File: packages/common-scripts/index.d.ts (hunk 2 above)
Comment (classicalliu) [reply packages/common-scripts/index.d.ts]: updated

File: packages/common-scripts/index.d.ts (hunk 3 above)
Comment (classicalliu) [reply packages/common-scripts/index.d.ts]: updated

File: packages/common-scripts/index.d.ts (hunk 4 above)
Comment (classicalliu) [reply packages/common-scripts/index.d.ts]: updated

File: packages/common-scripts/lib/secp256k1_blake160.js (hunk 5 above)
Comment (classicalliu) [reply packages/common-scripts/lib/secp256k1_blake160.js]: updated

File: packages/common-scripts/lib/secp256k1_blake160_multisig.js (hunk 6 above)
Comment (classicalliu) [reply packages/common-scripts/lib/secp256k1_blake160_multisig.js]: Moved to `.d.ts`

File: packages/helpers/lib/index.js (hunk 7 above)
Comment (classicalliu) [reply packages/helpers/lib/index.js]: updated

**Comment by classicalliu** [comment]
`secp256k1Blake160Multisig.transfer` only need `fromAddress` or  `fromInfo`
If both are filled, will check if `fromAddress` and `fromInfo` match.
If only `fromInfo`, it works well.
If only `fromAddress`, it will throw error if no inputs from this address before, for witness will need  `fromInfo`.

**Comment by xxuejie** [comment]
> `secp256k1Blake160Multisig.transfer` only need `fromAddress` or `fromInfo`
> If both are filled, will check if `fromAddress` and `fromInfo` match.
> If only `fromInfo`, it works well.
> If only `fromAddress`, it will throw error if no inputs from this address before, for witness will need `fromInfo`.

I think we can define a FromInfo type like following:

```
type FromInfo = MultisigScript | Address
```

In pure JavaScript, `MultisigScript` is an object, `Address` is a string so it won't be hard to tell one from the other. This can ensure the user at least uses one of them, we can rely on typing instead of comments to regulate the API.

**Comment by classicalliu** [comment]
> > `secp256k1Blake160Multisig.transfer` only need `fromAddress` or `fromInfo`
> > If both are filled, will check if `fromAddress` and `fromInfo` match.
> > If only `fromInfo`, it works well.
> > If only `fromAddress`, it will throw error if no inputs from this address before, for witness will need `fromInfo`.
> 
> I think we can define a FromInfo type like following:
> 
> ```
> type FromInfo = MultisigScript | Address
> ```
> 
> In pure JavaScript, `MultisigScript` is an object, `Address` is a string so it won't be hard to tell one from the other. This can ensure the user at least uses one of them, we can rely on typing instead of comments to regulate the API.

updated

**Pull request merged**